sda/websocket.go

```diff
@@ -0,0 +1,365 @@
+package sda
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"strconv"
+
+	"time"
+
+	"reflect"
+	"strings"
+
+	"sync"
+
+	"github.com/dedis/cothority/log"
+	"github.com/dedis/cothority/network"
+	"github.com/dedis/protobuf"
+	"github.com/gorilla/websocket"
+	"gopkg.in/tylerb/graceful.v1"
+)
+
+// WebSocket handles incoming client-requests using the websocket
+// protocol. When making a new WebSocket, it will listen one port above the
+// ServerIdentity-port-#.
+// The websocket protocol has been chosen as smallest common denominator
+// for languages including JavaScript.
+type WebSocket struct {
+	services  map[string]Service
+	server    *graceful.Server
+	mux       *http.ServeMux
+	startstop chan bool
+	started   bool
+	sync.Mutex
+}
+
+const (
+	// WebSocketErrorPathNotFound indicates the path has not been registered
+	WebSocketErrorPathNotFound = 4000 + iota
+	// WebSocketErrorProtobufDecode indicates an error in decoding the protobuf-packet
+	WebSocketErrorProtobufDecode
+	// WebSocketErrorProtobufEncode indicates an error in encoding the return packet
+	WebSocketErrorProtobufEncode
+	// WebSocketErrorInvalidErrorCode indicates the service returned
+	// an invalid error-code
+	WebSocketErrorInvalidErrorCode
+	// WebSocketErrorRead indicates that there has been a problem on reception
+	WebSocketErrorRead
+)
+
+// NewWebSocket opens a webservice-listener one port above the given
+// ServerIdentity.
+func NewWebSocket(si *network.ServerIdentity) *WebSocket {
+	w := &WebSocket{
+		services:  make(map[string]Service),
+		startstop: make(chan bool),
+	}
+	webHost, err := getWebHost(si)
+	log.ErrFatal(err)
+	w.mux = http.NewServeMux()
+	w.server = &graceful.Server{
+		Timeout: 100 * time.Millisecond,
+		Server: &http.Server{
+			Addr:    webHost,
+			Handler: w.mux,
+		},
+	}
+	return w
+}
+
+// Start listening on the port.
+func (w *WebSocket) Start() {
+	w.Lock()
+	w.started = true
+	w.Unlock()
+	log.Lvl3("Starting to listen on", w.server.Server.Addr)
+	go func() {
+		w.server.ListenAndServe()
+		w.startstop <- false
+	}()
+	w.startstop <- true
+}
+
+// RegisterService stores a service to the given path. All requests to that
+// path and it's sub-endpoints will be forwarded to ProcessClientRequest.
+func (w *WebSocket) RegisterService(service string, s Service) error {
+	w.services[service] = s
+	h := &wsHandler{
+		service:     s,
+		serviceName: service,
+	}
+	w.mux.Handle(fmt.Sprintf("/%s/", service), h)
+	return nil
+}
+
+// Stop the websocket and free the port.
+func (w *WebSocket) Stop() {
+	w.Lock()
+	defer w.Unlock()
+	if !w.started {
+		return
+	}
+	log.Lvl3("Stopping", w.server.Server.Addr)
+	<-w.startstop
+	w.server.Stop(100 * time.Millisecond)
+	<-w.startstop
+	w.started = false
+}
+
+// Client is a struct used to communicate with a remote Service running on a
+// sda.Conode. Using Send it can connect to multiple remote Conodes.
+type Client struct {
```

Comment:
Since `Client` might be the only public part of this file, we might as well put it first or at least next to the `ClientError` stuff. It justs seems strange to see in the order `Websocket, Client, wshandler, clientError`.